server-controller/hooks_test.go

```diff
@@ -29,104 +29,444 @@ import (
+	{
+		name:              "Deployment patch error",
+		inferenceEndpoint: inferenceEndpoint,
+		expected:          "ERROR during adding configDate label to deployment",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClient{patchError: errorNew},
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:              "Service creation error",
+		inferenceEndpoint: inferenceEndpoint,
+		expected:          "ERROR during service creation",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClient{createError: errorNew},
+			ingressClientError:    mockClientNoError,
+		},
+	},
+	{
+		name:              "Ingress creation error",
+		inferenceEndpoint: inferenceEndpoint,
+		expected:          "ERROR during ingress creation",
+		clientErrors: clientError{
+			configMapError:        mockClientNoError,
+			deploymentClientError: mockClientNoError,
+			serviceClientError:    mockClientNoError,
+			ingressClientError:    mockClient{createError: errorNew},
+		},
+	},
+}
+
+func testAdd(infer crv1.InferenceEndpoint, errs clientError) string {
+	var buf bytes.Buffer
+	log.SetOutput(&buf)
+	defer func() {
+		log.SetOutput(os.Stderr)
+	}()
+
+	updateMap := createTemplateClients(errs)
+	hooks := serverHooks{updateMap}
+	inferenceEndpoint := infer
+	hooks.Add(&inferenceEndpoint)
+	return buf.String()
+}
+
+var inferenceEndpointsTestUpdate = []struct {
+	name         string
+	oldServer    crv1.InferenceEndpoint
+	newServer    crv1.InferenceEndpoint
+	expected     string
+	clientErrors clientError
+}{
+	{
+		name:      "No such template",
+		oldServer: inferenceEndpoint,
+		newServer: crv1.InferenceEndpoint{
 			TypeMeta:   metav1.TypeMeta{},
 			ObjectMeta: metav1.ObjectMeta{},
-			Spec:       crv1.InferenceEndpointSpec{},
-			Status:     crv1.InferenceEndpointStatus{},
-		},
+			Spec:       crv1.InferenceEndpointSpec{EndpointName: "test", TemplateName: "not_exist"},
+			Status:     crv1.InferenceEndpointStatus{}},
 		expected:     "There is no such template",
-		clientErrors: clientErr{nil, nil, nil, nil}},
+		clientErrors: nilClientError,
+	},
 	{
-		name: "Success to create template",
-		inferenceEndpoint: crv1.InferenceEndpoint{
+		name:      "Updating succeeded",
+		oldServer: inferenceEndpoint,
+		newServer: crv1.InferenceEndpoint{
 			TypeMeta:   metav1.TypeMeta{},
 			ObjectMeta: metav1.ObjectMeta{},
-			Spec:       crv1.InferenceEndpointSpec{EndpointName: "test", TemplateName: "test"},
+			Spec:       crv1.InferenceEndpointSpec{EndpointName: "test", TemplateName: "exist"},
 			Status:     crv1.InferenceEndpointStatus{}},
-		expected:     "created successfully",
-		clientErrors: clientErr{nil, nil, nil, nil}},
+		expected:     "Updating succeeded for the server",
+		clientErrors: nilClientError,
+	},
+	{
+		name:         "No changes detected",
```

Comment:
During this review i see that we should change log in this case, because we don't have anything to patch in deployment and ingress, but probably something different have to be updated, so this message is confusing.
https://github.com/IntelAI/inference-model-manager/blob/master/server-controller/hooks.go#L233